parser/impl/parser.cpp

```diff
@@ -469,17 +469,201 @@ sPtr<ast::IfStatement> Parser::parseIfStatement() {
+    return parseBinaryExpression(0);
+}
+
+int Parser::opPrec(TokenType token) {
+    switch (token) {
+    case TokenType::Or:  // or
+    case TokenType::Xor: // xor
+        return 1;
+    case TokenType::And: // and
+        return 2;
+    case TokenType::Eq:      // =
+    case TokenType::Neq:     // /=
+    case TokenType::Less:    // <
+    case TokenType::Leq:     // <=
+    case TokenType::Greater: // >
+    case TokenType::Geq:     // >=
+        return 3;
+    case TokenType::Add: // +
+    case TokenType::Sub: // -
+        return 4;
+    case TokenType::Mul: // *
+    case TokenType::Div: // /
+    case TokenType::Mod: // %
+        return 5;
+    case TokenType::Not:       // not
+    case TokenType::OpenParen: // ()
+        return 6;
+    case TokenType::Dot:       // .
+    case TokenType::OpenBrack: // []
+        return 7;
+    default:
+        return -1; // unknown op
+    }
+}
+
+bool Parser::isPrimitive(TokenType token) {
+    return token == TokenType::Identifier || token == TokenType::Int ||
+           token == TokenType::Real || token == TokenType::True ||
+           token == TokenType::False;
+}
+
+sPtr<ast::Expression> Parser::parseUnaryExpression() {
+    Token tok = m_lexer.Peek();
+    ast::Expression expr;
+    if (opPrec(tok.type) >= 0) {
+        if (tok.type == TokenType::Not || tok.type == TokenType::Sub ||
+            tok.type == TokenType::Add) {
+
+            tok = m_lexer.Next();
+            expr.operation = tok;
+            expr.operand1 = parseUnaryExpression();
+            return std::make_shared<ast::Expression>(expr);
+
+        } else if (tok.type == TokenType::OpenParen) {
+            tok = m_lexer.Next();
+            sPtr<ast::Expression> expr = parseBinaryExpression(0);
+            if (m_lexer.Peek().type == TokenType::CloseParen) {
+                tok = m_lexer.Next();
+                return expr;
+            }
+            // error
+            m_errors.push_back(Error{
+                .pos = tok.pos,
+                .message = "Expected to find ')'",
+            });
+            return nullptr;
+        } else {
+            // throw error
+            m_errors.push_back(Error{
+                .pos = tok.pos,
+                .message = "Expected to find unary operator",
+            });
+            return nullptr;
+        }
+    } else if (isPrimitive(tok.type)) {
+        tok = m_lexer.Next();
+        // check if More Then Identifier
+        if (tok.type == TokenType::Identifier) {
+
+            Token t = m_lexer.Peek();
+            if (t.type == TokenType::OpenParen) {
+                // parse routineCall:
+                ast::RoutineCall rc;
+                rc.routine = tok; // save the function name
+                expr.operation = t;
+                do {
+                    t = m_lexer.Next(); // first call-read '(', others - ','
+                    // read expression
+                    sPtr<ast::Expression> e = parseBinaryExpression(0);
```

Comment:
It would look a bit nicer if this was just `parseExpression` to avoid tight coupling to the particular function